### PR TITLE
45 to be merged

### PR DESCRIPTION
**Pull request by jisie**
This a branch made off of cohort-45 branch so that the master branch can be updated.

### CODE REVIEW

**Comment by stevebrownlee** [comment]
:shipit:

File: book-4-nashville-kennels/chapters/AUTHENTICATION.md [hunk 1]
```diff
@@ -28,33 +27,39 @@ What determines if a user has authenticated? It's the `kennel_customer` key that
...
-                return <Redirect to="/login" />
-            }
-        }} />
-
-        <Route path="/login" render={props => <Login {...props} />} />
-        <Route path="/register" render={props => <Register {...props} />} />
-    </>
-)
+  <>
+    <Route
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/AUTHENTICATION.md:40]: It seems a little weird to use a `render` prop here and to not have a `path` prop

File: book-4-nashville-kennels/chapters/DATA_PROVIDER.md [hunk 2]
```diff
@@ -1,34 +1,34 @@
 # Data Providers in React
 
-Your data providers have so far been fairly simple modules. They have a module-scoped data structure that is the raw data for a resource you are using in the application. They also include a few methods to get, create, edit, or delete data from the API state.
+Your data providers have so far been fairly simple modules. They have a module-scoped data structure that contains the raw data for a resource used within the application. They also include a few methods: get, create, edit, or delete data from the API.
 
-Here's some example vanilla JavaScript code for a data provider.
+Here's an example vanilla JavaScript code for a data provider.
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/DATA_PROVIDER.md:5]: "Here's an example vanilla JavaScript **module** for a data provider."

File: book-4-nashville-kennels/chapters/DATA_PROVIDER.md [hunk 3]
```diff
@@ -95,63 +90,55 @@ export const LocationProvider = (props) => {
...
-You need to define a single property for each provider that you define in your system. This is because the components that use the data must be defined as children components _(more about this in the next chapter)_, and React will send an object to each component. One of the properties on that object will be `children`, which contains the child elements.
-
-### Defining Component State
+You define a single property for each provider defined in your system. This is because the components that uses the data must be defined as children components _(more about this in the next chapter)_, and React will send an object to each component. One of the properties on that object will be `children`, which contains the child elements.
 
-Next, you will use the State hook to define a variable that holds the state of the component, **and** a function to be used to modify that state.
+### Defining State Within a Component
 
-Yes, this is weird, and yes, it's hard to understand when you first see it.
+Next, you will use the State hook to define a variable that holds the state of the component, **and** a function that updates it.
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/DATA_PROVIDER.md:125]: "...you will use the `useState()` hook..."

File: book-4-nashville-kennels/chapters/DYNAMIC_ROUTING.md [hunk 4]
```diff
@@ -2,26 +2,28 @@
...
+import { useParams } from "react-router-dom"
+import "./Animal.css"
 
 
-export const EmployeeDetail = (props) => {
-    const { animals, getAnimals } = useContext(AnimalContext)
+export const AnimalDetail = (props) => {
+    const { animals, getOneAnimal } = useContext(AnimalContext)
     const { locations, getLocations } = useContext(LocationContext)
     const { employees, getEmployees } = useContext(EmployeeContext)
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/DYNAMIC_ROUTING.md:29]: These lines should be removed.

File: book-4-nashville-kennels/chapters/EDIT.md [hunk 5]
```diff
@@ -1,219 +1,223 @@
...
 import { LocationContext } from "../location/LocationProvider"
+import { AnimalContext } from "../animal/AnimalProvider"
+import { CustomerContext } from "../customer/CustomerProvider"
+import "./Animal.css"
+import { useHistory, useParams } from 'react-router-dom';
 
-
-export const AnimalForm = (props) => {
-    // Use the required context providers for data
+export const AnimalForm = () => {
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/EDIT.md:99]: As we discussed, this the Add and Edit forms should be different components.

File: book-4-nashville-kennels/chapters/FORMS_USEREF.md [hunk 6]
```diff
@@ -1,183 +1,183 @@
...
 ## Goal
 
-In this chapter, you are going to write a component whose responsibility is to create a new employee record and assign that employee to a location.
+In this chapter, you are going to write a component whose responsibility is to create a new animal record with a customer and location.
+
+**You will need to have completed a `LocationProvider` and `CustomerProvider`** before this chapter.
 
 ## New Concepts
 
 * useRef()
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/FORMS_USEREF.md:11]: As we discussed, we should use controlled components whenever possible.

File: book-4-nashville-kennels/chapters/FORMS_USEREF.md [hunk 7]
```diff
@@ -1,183 +1,183 @@
...
+          <button className="btn btn-primary"
+            onClick={handleClickNewAnimal}>
+            Save Animal
+          </button>
+      </form>
     )
 }
 ```
 
+<<<<<<< HEAD
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/FORMS_USEREF.md:178]: Merge issue

File: book-4-nashville-kennels/chapters/FORMS_USEREF.md [hunk 8]
```diff
@@ -205,25 +205,46 @@ This is because you imported the **`LocationProvider`** component and used the C
...
-You need to use the `history.push()` method again to take the user to a new view. Reference the code above to see how to do the following two tasks.
+### Create an EmployeeForm component.
+* Create a route in ApplicationViews for /employees/create that renders an EmployeeForm.
+* Add a button to the employee list labeled, "New Employee".
+* When the button is clicked, show the employee form by using history.push() to change the route.
+* The employee form should include an input for the name and a dropdown for the location.
+* On `Save`, create a new employee object and POST it to the API. The employee object should include the locationId as a foreign key.
+* Once the employee is saved, re-route the user to the list of employees.
 
+<<<<<<< HEAD
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/FORMS_USEREF.md:244]: Merge issue

File: book-4-nashville-kennels/chapters/LIST_USECONTEXT.md [hunk 9]
```diff
@@ -1,161 +1,139 @@
...
+      }
+    </div>
+  )
 }
 ```
 
 ### Breaking it Down
 
-In the above component, you will notice a new hook being imported from React - the Context hook. This hook allows you to use data structures and functions that a parent provider component exposes. In the previous chapter, your **`LocationProvider`** component exposed the following two elements.
+In the above component, you will notice a 2 new hooks: `useContext` and `useEffect`.
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/LIST_USECONTEXT.md:44]: two

File: book-4-nashville-kennels/chapters/LIST_USECONTEXT.md [hunk 10]
```diff
@@ -1,161 +1,139 @@
...
-        <h2>Nashville Kennels</h2>
-        <small>Loving care when you're not there.</small>
-
-        <h2>Locations</h2>
-        <LocationProvider>
-            <LocationList />
-        </LocationProvider>
-    </>
-)
+<AnimalProvider>
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/LIST_USECONTEXT.md:105]: Are we going to leave these providers around the routes, or use global providers in `App.js`?

File: book-4-nashville-kennels/chapters/ROUTING.md [hunk 11]
```diff
@@ -78,44 +81,60 @@ Notice the use of the `<Link/>` component. This comes from the React Router pack
...
 * `/animals`
 * `/customers`
 * `/employees`
 
 In the **`ApplicationViews`** component, you will define how your application will respond when the URL matches each of those patterns. When a user clicks on one of the hyperlinks in the navigation bar, this code dictates which component should be rendered.
 
+**For this example, convert the static content of the `Kennel` component into a `Home` component.
+
+> ##### `src/components/Home.js`
+```
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/ROUTING.md:94]: add `jsx`

File: book-4-nashville-kennels/chapters/ROUTING.md [hunk 12]
```diff
@@ -78,44 +81,60 @@ Notice the use of the `<Link/>` component. This comes from the React Router pack
...
+            <div>500 Puppy Way</div>
+        </address>
+        <PropsAndState yourName={"Brenda"} />
+    </>
+)
+```
+
 > ##### `src/components/ApplicationViews.js`
 
 ```js
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/ROUTING.md:114]: jsx

File: book-4-nashville-kennels/chapters/scripts/auth.sh [hunk 13]
```diff
@@ -4,66 +4,58 @@ set -u
...
-                    }
-                    else {
-                        conflictDialog.current.showModal()
-                    }
-                })
-        } else {
-            passwordDialog.current.showModal()
-        }
+                    })
+                        .then(_ => _.json())
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/scripts/auth.sh:41]: replace `_` with `res` or `resp`

File: book-4-nashville-kennels/chapters/scripts/auth.sh [hunk 14]
```diff
@@ -4,66 +4,58 @@ set -u
...
+                        .then(createdUser => {
+                            if (createdUser.hasOwnProperty("id")) {
```
Comment (askingalot) [inline book-4-nashville-kennels/chapters/scripts/auth.sh:43]: I don't understand this `if`

**Review by askingalot — CHANGES_REQUESTED** [review]
I've left a few comments. Some of them are opinions, Some of them are probably better resolved in future changes and shouldn't stop the merge, but a couple should be addressed here. I think those will be obvious, but I'm happy to talk about it.

**Comment by JoeShep** [comment]
Thanks for the thorough review, @askingalot. No further comments from me at this time.

**Pull request merged**